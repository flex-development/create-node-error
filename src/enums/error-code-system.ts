/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
