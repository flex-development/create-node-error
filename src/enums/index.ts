/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
