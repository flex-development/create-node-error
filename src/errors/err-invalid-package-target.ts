/**
 * @file Errors - ERR_INVALID_PACKAGE_TARGET
 * @module errnode/errors/ERR_INVALID_PACKAGE_TARGET
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1488-L1503
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import { sep } from '@flex-development/pathe'
import { DOT } from '@flex-development/tutils'
import { ok } from 'devlop'

/**
 * `ERR_INVALID_PACKAGE_TARGET` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_invalid_package_target
 *
 * @extends {NodeError<codes.ERR_INVALID_PACKAGE_TARGET>}
 */
interface ErrInvalidPackageTarget
  extends NodeError<codes.ERR_INVALID_PACKAGE_TARGET> {}

/**
 * `ERR_INVALID_PACKAGE_TARGET` message arguments.
 */
type Args = [
  dir: string,
  subpath: string,
  target: unknown,
  internal?: boolean | null | undefined,
  base?: URL | string | null | undefined
]

/**
 * `ERR_INVALID_PACKAGE_TARGET` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrInvalidPackageTarget}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrInvalidPackageTarget,Args>}
 */
interface ErrInvalidPackageTargetConstructor
  extends NodeErrorConstructor<ErrInvalidPackageTarget, Args> {
  /**
   * Create a new `ERR_INVALID_PACKAGE_TARGET` error.
   *
   * @see {@linkcode ErrInvalidPackageTarget}
   *
   * @param {string} dir
   *  Package directory module id
   * @param {string} subpath
   *  Package subpath
   * @param {unknown} target
   *  Invalid package target
   * @param {boolean | null | undefined} [internal]
   *  Internal `subpath`?
   * @param {URL | string | null | undefined} [base]
   *  Parent module id
   * @return {ErrInvalidPackageTarget}
   */
  new (
    dir: string,
    subpath: string,
    target: unknown,
    internal?: boolean | null | undefined,
    base?: URL | string | null | undefined
  ): ErrInvalidPackageTarget
}

/**
 * `ERR_INVALID_PACKAGE_TARGET` model.
 *
 * Thrown when a `package.json` [`"exports"`][1] or [`"imports"`][2] field
 * contains an invalid target mapping value for the attempted module resolution.
 *
 * [1]: https://nodejs.org/api/packages.html#exports
 * [2]: https://nodejs.org/api/packages.html#imports
 *
 * @see {@linkcode ErrInvalidPackageTargetConstructor}
 *
 * @type {ErrInvalidPackageTargetConstructor}
 *
 * @class
 */
const ERR_INVALID_PACKAGE_TARGET: ErrInvalidPackageTargetConstructor = E(
  codes.ERR_INVALID_PACKAGE_TARGET,
  Error,
  /**
   * @param {string} dir
   *  Package directory module id
   * @param {string} subpath
   *  Package subpath
   * @param {unknown} target
   *  Invalid package target
   * @param {boolean | null | undefined} [internal]
   *  Internal `subpath`?
   * @param {URL | string | null | undefined} [base]
   *  Parent module id
   * @return {string}
   *  Error message
   */
  function message(
    dir: string,
    subpath: string,
    target: unknown,
    internal: boolean | null | undefined = false,
    base: URL | string | null | undefined = null
  ): string {
    ok(dir.endsWith('/'), 'expected `dir` to end with trailing slash ("/")')

    /**
     * Relative error?
     *
     * @const {boolean} relativeError
     */
    const relativeError: boolean = typeof target === 'string' &&
      internal !== true &&
      target.length > 0 &&
      !target.startsWith(`${DOT}/`)

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = ''

    if (subpath === DOT) {
      ok(internal !== true, 'expected `internal` to not be `true`')

      message += `Invalid "exports" main target ${JSON.stringify(target)}` +
        ` defined in the package config ${dir}package.json`
    } else {
      message = `Invalid "${internal ? 'imports' : 'exports'}" target` +
        ` ${JSON.stringify(target)} defined for '${subpath}' in the package` +
        ` config ${dir}package.json`
    }

    if (base !== null) message += ` imported from ${String(base)}`
    if (relativeError) message += `; targets must start with "${DOT}${sep}"`

    return message
  }
)

export {
  ERR_INVALID_PACKAGE_TARGET as default,
  type ErrInvalidPackageTarget,
  type Args as ErrInvalidPackageTargetArgs,
  type ErrInvalidPackageTargetConstructor
}
