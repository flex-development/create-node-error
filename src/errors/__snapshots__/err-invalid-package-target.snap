// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_INVALID_PACKAGE_TARGET > should build message from parameters (0) 1`] = `Invalid "exports" main target "dist/index.mjs" defined in the package config \${process.cwd()}/node_modules/@flex-development/pathe/package.json imported from \${process.cwd()}/loader.mjs; targets must start with "./"`;

exports[`unit:errors/ERR_INVALID_PACKAGE_TARGET > should build message from parameters (1) 1`] = `Invalid "exports" target "../package.json" defined for './package.json' in the package config \${process.cwd()}/node_modules/@flex-development/pathe/package.json imported from \${process.cwd()}/src/errors/__tests__/err-invalid-package-target.spec.ts; targets must start with "./"`;

exports[`unit:errors/ERR_INVALID_PACKAGE_TARGET > should build message from parameters (2) 1`] = `Invalid "imports" target "src/index.ts" defined for '#src' in the package config \${process.cwd()}/package.json imported from \${process.cwd()}/src/errors/__tests__/err-invalid-package-target.spec.ts`;
